src/Native/System.Native/pal_networkstatistics.cpp

```diff
@@ -412,4 +414,114 @@ extern "C" int32_t GetActiveUdpListeners(IPEndPointInfo* infos, int32_t* infoCou
     return 0;
 }
 
+extern "C" int32_t GetNativeIPInterfaceStatistics(char* interfaceName, NativeIPInterfaceStatistics* retStats)
+{
+    assert(interfaceName != nullptr && retStats != nullptr);
+    unsigned int interfaceIndex = if_nametoindex(interfaceName);
+    if (interfaceIndex == 0)
+    {
+        // An invalid interface name was given (doesn't exist).
+        return -1;
+    }
+
+    int statisticsMib[] =
+    {
+        CTL_NET,
+        PF_ROUTE,
+        0,
+        0,
+        NET_RT_IFLIST2,
+        0
+    };
+    
+    size_t len;
+    // Get estimated data length
+    if (sysctl(statisticsMib, 6, nullptr, &len, nullptr, 0) == -1)
+    {
+        memset(retStats, 0, sizeof(NativeIPInterfaceStatistics));
+        return -1;
+    }
+
+    uint8_t* buffer = new uint8_t[len];
```

Comment:
I think these were pretty much all leaking except in error cases. I've gone through and checked that all the buffers are being freed in all code paths.